### ISSUE TITLE
Federated Learning API

### ISSUE THREAD

**Issue by mortendahl** [tf-encrypted/tf-encrypted#670]
## Goals

- reduce burden on users
  - align with TF as far as possible by eg matching their API
  - align with TF Federated where possible
  - find common ground across TF Encrypted, and minimise concepts and terminology specific to federated learning
  - reduce redundancy and ambiguity; make sure every construct has a reason

- ensure flexibility to experiment
  - models, aggregation strategies, and cryptographic techniques

- keep *what* separated from *how* to lower complexity and increase reuse

  - allow higher-order ops such as `Dense` and `ReLU` to sometimes let the how depend on the what

### TF Distribute strategies

From [the guide](https://www.tensorflow.org/guide/distribute_strategy#using_tfdistributestrategy_with_keras):

> The only things that need to change in a user's program are: (1) Create an instance of the appropriate `tf.distribute.Strategy` and (2) Move the creation and compiling of Keras model inside `strategy.scope`.

> `strategy.scope()` indicated which parts of the code to run distributed. Creating a model inside this scope allows us to create mirrored variables instead of regular variables. Compiling under the scope allows us to know that the user intends to train this model using this strategy. Once this is setup, you can fit your model like you would normally [i.e. outside scope].

```python
mirrored_strategy = tf.distribute.MirroredStrategy()
with mirrored_strategy.scope():
  model = tf.keras.Sequential([tf.keras.layers.Dense(1, input_shape=(1,))])
  model.compile(loss='mse', optimizer='sgd')

dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(100).batch(10)
model.fit(dataset, epochs=2)
model.evaluate(dataset)
```

### TF Federated

TODO

## Terminology

**Functionality** and **Protocol**: the former is basically used as a function from and to local tensors, while the latter as a more general means to specify how functionalities are to be computed using e.g. cryptographic techniques. As such, protocols are the only one of the two that are used as context handlers. Roughly follows UC terminology although functionalities are not intended to be used as sub-protocols.

## Suggested API

```python
# specify the players involved
model_owner = tfe.Player('model_owner')
data_owners = [
    tfe.Player('data_owner_0'),
    tfe.Player('data_owner_1'),
    tfe.Player('data_owner_2'),
]

# build data pipeline on each data owner;
# this would likely be an unique function per owner
data_sources = [
    build_data_pipeline(data_owner)
    for data_owner in data_owners
]

# use fast, non-resilient, secure aggregation based on additive secret sharing
aggregation = tfe.functionalities.AdditiveSecureAverage

# ... alternatively we could have instantiated it explicitly,
# resulting in exactly the same thing
aggregation = tfe.functionalities.AdditiveSecureAverage(
    compute_players=data_owners,
    output_receiver=model_owner)

# ... or we could have unrolled its (simplified) implementation
def aggregation(plaintext_grads):
  pond = tfe.protocols.Pond(data_owners)
  with pond:
    grads = [
        tfe.define_private_input(grad, owner)
        for grad, owner in zip(plaintext_grads, data_owners)
    ]
    aggregated_grad = tfe.add_n(grads) / len(grads)
    return tfe.reveal(aggregated_grad, model_owner)

# initialising the federated protocol with the model owner in order
# to specify where the reference weights of the model should live
# and where updates should happen
federated = tfe.protocol.FederatedLearning(model_owner, aggregation)

with federated:

  # creating the model here essentially creates a `model_fn` function (not unlike
  # what is done in TFF) that will be executed on both model and data owners
  model = tfe.keras.Sequential()
  model.add(tfe.keras.Dense())
  model.add(tfe.keras.ReLU())

  # compiling in this context implies that we want the model updates to be
  # applied on the model owner; compiling outside the context would cause
  # an error due to wrong locality between tensors
  model.compile(
      optimizer=tf.train.AdamOptimizer(0.001),
      loss='categorical_crossentropy',
      metrics=['accuracy'])

# fitting can be done anywhere following ordinary TF; 
model.fit(data_sources, epochs=2)
model.evaluate(data_sources)
```

## Broad Fit

```python
client = tfe.Player()
server = tfe.Player()

gazelle = tfe.protocols.Gazelle(client, server)

with gazelle:
  model = tfe.keras.Sequential()
  model.add(tfe.keras.Dense())
  model.add(tfe.keras.ReLU())

y = model.predict(x)

#
# which unrolls to the following
#

# instantiate HE scheme
he_scheme = tfe.protocols.HE(
    key_owner=client,
    evaluator=server)

# instantiate GC scheme
gc_scheme = tfe.protocols.GC(
    garbler=client,
    evaluator=server)

def fn(plaintext_x):

  # create a model that lives on the server,
  # i.e. all weights are stored there by default
  with tf.device(server):
    model = tfe.keras.Sequential()
  
  x = plaintext_x

  with he_scheme:
    x = tfe.define_private_input(x, client)
    x = tfe.keras.Dense(x)
    x = tfe.define_output(x, client)

  with gc_scheme:
    x = tfe.define_private_input(x, client)
    x = tfe.keras.ReLU(x)
    x = tfe.define_output(x, client)

  return x

y = fn(x)
```

```python
model_trainer = tfe.Player()
data_owners = [tfe.Player(), ..]

enclave = tfe.protocols.Enclave(enclave_owner)

with enclave:
  model = tfe.keras.Sequential()
  model.add(tfe.keras.Dense())
  model.add(tfe.keras.ReLU())

dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(100).batch(10)

with enclave.scope():
  model.fit(dataset, epochs=2)
  model.evaluate(dataset)
```

**Comment by jvmancuso** [comment]
Spoke with @mortendahl briefly and we agreed that much of the code beneath this API can be written as a custom tf.distribute.Strategy. Roughly speaking, the `aggregate` function above could be written in the language of these Strategies. This should simplify how this interface works with the Keras and Estimator APIs.

**Comment by mortendahl** [comment]
Note to self: find a way where the federated protocol simply wraps an existing model on the model owner?

**Comment by mortendahl** [comment]
Discussion has moved into a RFC: https://github.com/tf-encrypted/rfcs/pull/1

**Comment by mortendahl** [comment]
@yanndupis @jvmancuso the more I look into this the more I think we should simply stick with the FL example we have, but upgrade it to use TF Keras and TF 2.0.

I'll keep working on the details in integrating with TF Federated and distribution strategies but worried the time will be short and not worth rushing.

WDYT?

**Comment by yanndupis** [comment]
@mortendahl - happy to explore other alternatives. In general, I think it's important that's easy to use and we can support arbitrary Keras models. So maybe we just need to improve the abstraction and automate some of the stuff in the example. For example [here], we would like to avoid defining manually the model weights. 

It would be nice to kick the training process with `model.fit(data_sources, epochs=10)` and see the training process. Also I think we would like to have some utils to distribute the data among several data owners.

**Comment by jvmancuso** [comment]
>@yanndupis @jvmancuso the more I look into this the more I think we should simply stick with the FL example we have, but upgrade it to use TF Keras and TF 2.0.

Yeah, I agree. Let's stick with the current abstraction and add some syntactic sugar, like what @yanndupis referred to.

**Comment by justin1121** [comment]
> It would be nice to kick the training process with model.fit(data_sources, epochs=10) and see the training process. 

@yanndupis how do you think we could accomplish the above? Would we create a keras model wrapper? Or be able to make use of tfe.keras somehow?

> Also I think we would like to have some utils to distribute the data among several data owners.

Can you provide some more details about this also? Not 100% sure what you mean.

**Comment by yanndupis** [comment]
> Can you provide some more details about this also? Not 100% sure what you mean.

In tf federated, they have this concept of [tff.simulation.datasets.emnist.load_data](https://www.tensorflow.org/federated/tutorials/federated_learning_for_image_classification). It constructs a `tf.data.Dataset` object for each data owner. Currently [here](https://github.com/tf-encrypted/tf-encrypted/blob/259087e444c536db1747c1784765bf93f9464491/examples/federated-learning/run.py#L175), the number of data owner and tfrecord file is hard coded. If we want be able to easily experiment, it would be nice to have a method which takes data as a numpy format (or even tfrecord) and number of players. Then the data would be distributed evenly among these players. 

For `model.fit`, I will put more thoughts into it tomorrow. It could be just wrapping up [this section](https://github.com/tf-encrypted/tf-encrypted/blob/259087e444c536db1747c1784765bf93f9464491/examples/federated-learning/run.py#L196) in a `.fit`. I also like the API above suggested by @mortendahl, we just need to assess if it's too much work for our timeline.

**Comment by mortendahl** [comment]
> For example [here], we would like to avoid defining manually the model weights.

Yes, fully agree.

> It would be nice to kick the training process with `model.fit(data_sources, epochs=10)` and see the training process.

Following the current abstractions in the example, it may make more sense to take the data owners as input instead of the data sources.

>  it would be nice to have a method which takes data as a numpy format (or even tfrecord) and number of players. Then the data would be distributed evenly among these players.

This seems useful indeed, but let's make sure to clearly mark it as something that's for simulation use only.

**Comment by justin1121** [comment]
This is being worked on [here](https://github.com/tf-encrypted/tf-encrypted/pull/684). Feedback welcome!

**Comment by justin1121** [comment]
In https://github.com/tf-encrypted/tf-encrypted/pull/695 we added the flag to support simulating the splitting the data into the parties. It currently only works with local computations we should see if we can get this to work with remote computations with either `tf.device` or some other way. See this thread: https://github.com/tf-encrypted/tf-encrypted/pull/695#discussion_r335050263
